fluent-bundle/src/builtins.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @overview
+ *
+ * The FTL resolver ships with a number of functions built-in.
+ *
+ * Each function take two arguments:
+ *   - args - an array of positional args
+ *   - opts - an object of key-value args
+ *
+ * Arguments to functions are guaranteed to already be instances of
+ * `FluentType`.  Functions must return `FluentType` objects as well.
+ */
+
+import {
+  FluentNone,
+  FluentNumber,
+  FluentDateTime,
+  FluentType
+} from "./types.js";
+
+function values(opts: Record<string, FluentType>) {
+  const unwrapped: Record<string, unknown> = {};
+  for (const [name, opt] of Object.entries(opts)) {
+    unwrapped[name] = opt.valueOf();
+  }
+  return unwrapped;
+}
+
+export function NUMBER(
+  args: Array<FluentType>,
+  opts: Record<string, FluentType>
+) {
+  if (args.length !== 1) {
+    throw new TypeError("NUMBER accepts exactly one argument");
+  }
+  let arg = args[0];
+
+  if (arg instanceof FluentNone) {
+    return new FluentNone(`NUMBER(${arg.valueOf()})`);
+  }
+
+  if (arg instanceof FluentNumber) {
+    return new FluentNumber(arg.valueOf(), { ...arg.opts, ...values(opts) });
+  }
+
+  throw new TypeError("Invalid argument to NUMBER");
+}
+
+export function DATETIME(
+  args: Array<FluentType>,
+  opts: Record<string, FluentType>
+) {
+  if (args.length !== 1) {
+    throw new TypeError("DATETIME accepts exactly one argument");
+  }
+  let arg = args[0];
+
+  if (arg instanceof FluentNone) {
+    return new FluentNone(`DATETIME(${arg.valueOf()})`);
+  }
+
+  if (arg instanceof FluentNumber) {
+    return new FluentDateTime(arg.valueOf(), { ...arg.opts, ...values(opts) });
```

Comment:
I'm having second thoughts about this.

In TypeScript, this works because [`Intl.DateTimeFormatOptions`](
https://github.com/microsoft/TypeScript/blob/2ab1b71fff3bf54d5150ed29ddc71217d0018ad5/lib/lib.es5.d.ts#L4273-L4287) happens to be assignable to [`Intl.NumberFormatOptions`](
https://github.com/microsoft/TypeScript/blob/2ab1b71fff3bf54d5150ed29ddc71217d0018ad5/lib/lib.es5.d.ts#L4236-L4247). I'm not sure if we should rely on this.

On a higher level, this is a spec question: what should be the outcome of `NUMBER($date)` and `DATETIME($number)`? I think it will be safer to not make assumptions about the answer right now and wait for the formatting spec to be finished.

I'll revert the `FluentDateTime extends FluentNumber` change.